cloudinit/sources/DataSourceCloudStack.py

```diff
@@ -93,6 +93,60 @@ def __init__(self, sys_cfg, distro, paths):
         self.metadata_address = "http://%s/" % (self.vr_addr,)
         self.cfg = {}
 
+    def _get_domainname(self):
+        """
+        Try obtaining a "domain-name" DHCP lease parameter:
+        - From systemd-networkd lease
+        - From dhclient lease
+        """
+        LOG.debug("Try obtaining domain name from networkd leases")
+        domainname = dhcp.networkd_get_option_from_leases("DOMAINNAME")
+        if domainname:
+            return domainname
+        LOG.debug(
+            "Could not obtain FQDN from networkd leases. "
+            "Falling back to ISC dhclient"
+        )
+
+        lease_file = dhcp.IscDhclient.get_latest_lease()
+        if not lease_file:
+            LOG.debug("Dhclient lease file wasn't found")
+            return None
+
+        for lease in dhcp.IscDhclient.parse_dhcp_lease_file(lease_file):
```

Comment:
Are we really okay with a stale lease here? Would a stale lease be better than a non-fqdn? I'm not sure it is. Why not just check the most recent lease here?

Also, the most recent lease looks to be in [the last index of the array](https://github.com/canonical/cloud-init/blob/4d6b96d8bdc8b1169600790f3182c7e1b02eb68e/cloudinit/net/ephemeral.py#L293), so even if stale data is okay for some reason, we should really be checking the most recent first.